jumpscale/sals/marketplace/chatflow.py

```diff
@@ -29,3 +37,147 @@ def _validate_user(self):
             raise StopChatFlow(
                 f"You must accept terms and conditions before using this solution. please head towards the main page to read our terms"
             )
+
+    def _init_solution(self):
+        self._validate_user()
+        self.solution_id = uuid.uuid4().hex
+        self.solution_metadata = {}
+        self.solution_metadata["owner"] = self.user_info()["username"]
+        self.threebot_name = j.data.text.removesuffix(self.user_info()["username"], ".3bot")
+        self.query = dict()
+
+    def _wgconf_show_check(self):
+        if hasattr(self, "wgconf"):
+            self.download_file(msg=f"<pre>{self.wgconf}</pre>", data=self.wgconf, filename="apps.conf", html=True)
+
+    def _get_pool(self):
+        available_farms = []
+        for farm_name in FARM_NAMES:
+            available, _, _, _, _ = deployer.check_farm_capacity(farm_name, currencies=[self.currency], **self.query)
+            if available:
+                available_farms.append(farm_name)
+
+        self.farm_name = random.choice(available_farms)
+
+        user_networks = solutions.list_network_solutions(self.solution_metadata["owner"])
+        networks_names = [n["Name"] for n in user_networks]
+        if "apps" in networks_names:
+            # old user
+            self.pool_info = deployer.create_solution_pool(
+                bot=self,
+                username=self.solution_metadata["owner"],
+                farm_name=self.farm_name,
+                expiration=self.expiration,
+                currency=self.currency,
+                **self.query,
+            )
+            result = deployer.wait_pool_payment(self, self.pool_info.reservation_id)
+            if not result:
+                raise StopChatFlow(f"Waiting for pool payment timedout. pool_id: {self.pool_info.reservation_id}")
+        else:
+            # new user
+            self.pool_info, self.wgconf = deployer.init_new_user(
+                bot=self,
+                username=self.solution_metadata["owner"],
+                farm_name=self.farm_name,
+                expiration=self.expiration,
+                currency=self.currency,
+                **self.query,
+            )
+
+        if not self.pool_info:
+            raise StopChatFlow(f"Failed to deploy solution {self.pool_info}")
+        self.pool_id = self.pool_info.reservation_id
+        return self.pool_id
+
+    def _deploy_network(self):
```

Comment:
most of these methods are required for apps chatflows only not all marketplace solutions. can we apps inherit from a different class which includes these stuff?

```python
class AppChatflow(MarketplaceChatflow):
```

Reply:
Added `MarketPlaceAppsChatflow` base class and use it in apps.
Thanks @m-motawea